### PR TITLE
Add implicit conversions to function registry

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by danking** [comment]
@tpoterba Rather than review, I made a few changes, mostly around using a right-biased Either from the [`cats`](http://typelevel.org/cats/) library.

I think there's still a bit more work to do (see the commented out messages). I suspect we might actually want an `LookupError` sealed trait with an alternate for "ambiguous" and "not found" so that the consumer (i.e. `AST.scala`) can produce custom messages if desired.

**Comment by tpoterba** [comment]
👍

File: src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala [hunk 1]
```diff
@@ -217,11 +274,7 @@ object FunctionRegistry {
   registerField("abs", { (x: Float) => x.abs })
   registerField("abs", { (x: Double) => x.abs })
 
-  registerField("signum", { (x: Int) => x.signum })
-  registerField("signum", { (x: Long) => x.signum })
-  registerField("signum", { (x: Float) => x.signum })
   registerField("signum", { (x: Double) => x.signum })
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala]: isn't signum always an int?

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala]:
```scala
scala> -3.signum
res0: Int = -1

scala> -3d.signum
res1: Int = -1
```

Comment (danking) [inline src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala]: Apparently it does.

**Pull request merged**